node/pkg/query/request.go

```diff
@@ -0,0 +1,470 @@
+package query
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"math"
+	"strings"
+
+	"github.com/certusone/wormhole/node/pkg/common"
+	gossipv1 "github.com/certusone/wormhole/node/pkg/proto/gossip/v1"
+	"github.com/wormhole-foundation/wormhole/sdk/vaa"
+
+	ethCommon "github.com/ethereum/go-ethereum/common"
+	ethCrypto "github.com/ethereum/go-ethereum/crypto"
+)
+
+// QueryRequest defines a cross chain query request to be submitted to the guardians.
+// It is the payload of the SignedQueryRequest gossip message.
+type QueryRequest struct {
+	Nonce           uint32
+	PerChainQueries []*PerChainQueryRequest
+}
+
+// PerChainQueryRequest represents a query request for a single chain.
+type PerChainQueryRequest struct {
+	// ChainId indicates which chain this query is destine for.
+	ChainId vaa.ChainID
+
+	// Query is the chain specific query data.
+	Query ChainSpecificQuery
+}
+
+// ChainSpecificQuery is the interface that must be implemented by a chain specific query.
+type ChainSpecificQuery interface {
+	Type() ChainSpecificQueryType
+	Marshal() ([]byte, error)
+	Unmarshal(data []byte) error
+	UnmarshalFromReader(reader *bytes.Reader) error
+	Validate() error
+}
+
+// ChainSpecificQueryType is used to interpret the data in a per chain query request.
+type ChainSpecificQueryType uint8
+
+// EthCallQueryRequestType is the type of an EVM eth_call query request.
+const EthCallQueryRequestType ChainSpecificQueryType = 1
+
+// EthCallQueryRequest implements ChainSpecificQuery for an EVM eth_call query request.
+type EthCallQueryRequest struct {
+	// BlockId identifies the block to be queried. It mus be a hex string starting with 0x. It may be a block number or a block hash.
+	BlockId string
+
+	// CallData is an array of specific queries to be performed on the specified block, in a single RPC call.
+	CallData []*EthCallData
+}
+
+// EthCallData specifies the parameters to a single EVM eth_call request.
+type EthCallData struct {
+	// To specifies the contract address to be queried.
+	To []byte
+
+	// Data is the ABI encoded parameters to the query.
+	Data []byte
+}
+
+const EvmContractAddressLength = 20
+
+// PerChainQueryInternal is an internal representation of a query request that is passed to the watcher.
+type PerChainQueryInternal struct {
+	RequestID  string
+	RequestIdx int
+	Request    *PerChainQueryRequest
+}
+
+// QueryRequestDigest returns the query signing prefix based on the environment.
+func QueryRequestDigest(env common.Environment, b []byte) ethCommon.Hash {
+	// TODO: should this use a different standard of signing messages, like https://eips.ethereum.org/EIPS/eip-712
```

Comment:
If this will eventually be not just on evm chains, keep it as is. If it is going to not be expanded, try to do the best idiomatic "evm way" of doing this, such as eip-712.

Reply:
Should be handled in #3423 - @bruce-riley maybe we remove this comment given the above.